### ISSUE TITLE
event collapsing multithreads issue

### ISSUE THREAD

**Issue by lingolingolin** [a-slide/NanopolishComp#22]
Hi @a-slide ,

Thanks for developing this useful program.
I am interested in the way you using multiprocessing process and [queue.](https://github.com/a-slide/NanopolishComp/blob/master/NanopolishComp/Eventalign_collapse.py#L128)
I wonder what will happen if ***split_reads*** is slower than ***process_read***? 
In another word, what will happen when  ***process_read*** consumes ***in_q*** contents faster than ***split_reads*** can put?
Thanks a lot.

**Comment by a-slide** [comment]
Hi @lingolingolin,
This is a typical IO bound / distributed computing problem. `split_reads` doesn't do much apart from reading the input file. `Eventalign_collapse` is implemented in a highly thread safe fashion, as a single reader, multiple workers, single writer.
The input file generated by Nanopolish enventalign is huge and in my experience reading it is the bottleneck, in particular on High Performance Computing clusters. Now, if you run it locally on a desktop computer (with a high speed SSD disk) IO might not be the limit any more. However given the very large size of the input file I assume that it make more sense to use it on a HPC.
Depending on the speed of your file system, it might not be necessary to allocate many threads for the workers. I normally use 20 on our HPC and 5 locally on my laptop.
I hope it make sense.
Best

**Comment by lingolingolin** [comment]
Hi @a-slide ,
Many thanks for your explanation, which makes sense to me. 
My concern/question arises from the premise that reader, worker, and writer all work simultaneously. Given reader is a single thread process, the multiple threads worker may empty the Queue object created by the reader before reader can finish its task. I wonder if this is going to lead the breaking down of worker and the whole program.
I am uncertain about how python multiprocessing manages multiprocessors and I haven't found a clue from its document.

**Comment by a-slide** [comment]
No worries,
The implementation of Queues I used are from the multiprocessing package and it is thread safe.
So if the queue is empty, the workers will just wait until a new element is added to the queue and stop when they find a poison pill (in this case when the find a "None" value). 
On the other hand if the queue are growing faster than the workers can cope with, I capped the queues size to 1000 elements to prevent memory overflow.
